sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,704 @@
+            let ctx = &mut BigNumContext::new().unwrap();
+            let mut num;
+            loop {
+                let mut hash = blake3::Hasher::digest(input.as_slice());
+                // Force it to be odd
+                hash[31] |= 1;
+                // Only need 256 bits just borrow the bottom 32 bytes
+                // There should be plenty of primes below 2^256
+                // and we want this to be reasonably fast
+                //num = BigNumber::from_bytes(&hash[32..]);
+                num = BigNum::from_slice(&hash).unwrap();
+                if num.is_prime(15, ctx).unwrap() {
+                    // msg!("num_bytes:{:?}",num.to_bytes().to_vec());
+                    // msg!("i:{}",i);
+                    break;
+                }
+                i += 1;
+                let i_bytes = i.to_be_bytes();
+                input[offset..end].clone_from_slice(&i_bytes[..]);
+            }
+            let res_bn = Box::new(num);
+            let rwptr = Box::into_raw(res_bn);
+            Self(rwptr as u64)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            let mut bignum_ptr = nonce.len() as u64;
+            unsafe {
+                sol_bignum_hash_to_prime(
+                    &u.0 as *const _ as *const u64,
+                    &a.0 as *const _ as *const u64,
+                    &z.0 as *const _ as *const u64,
+                    nonce as *const _ as *const u64,
+                    &mut bignum_ptr as *mut _ as *mut u64,
+                );
+            }
+            Self(bignum_ptr)
+        }
+    }
+
+    /// Log a `BigNum` from a program
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        crate::program_stubs::sol_log(&self.to_string());
+        #[cfg(target_arch = "bpf")]
+        {
+            unsafe { sol_log_bignum(&self.0 as *const _ as *const u64) };
+        }
+    }
+}
+
+impl fmt::Debug for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use openssl::bn::BigNum;
+            let braw: &BigNum = unsafe { &*(self.0 as *mut BigNum) };
+            write!(f, "{}", braw)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl fmt::Display for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use openssl::bn::BigNum;
+            let braw: &BigNum = unsafe { &*(self.0 as *mut BigNum) };
+            write!(f, "{}", braw)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl AsRef<u64> for BigNumber {
+    fn as_ref(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl From<u64> for BigNumber {
+    fn from(ptr: u64) -> Self {
+        Self(ptr)
+    }
+}
+
+/// Drop - removes the underlying BigNum
+impl Drop for BigNumber {
```

Comment:
This means that if someone hacks thier program they could force a leaked bignum in the runtime?